### ISSUE TITLE
Update to XDS v3 API

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#19885]
Envoy has the new v3 API. We need to see if anything will break us, or if it's just a version bump for us. Namely, virtual listeners.

**Comment by hzxuzhonghu** [comment]
When do we plan to update to v3?

**Comment by howardjohn** [comment]
@hzxuzhonghu I think it depends on if there are any major changes that will impact us or if its just replace v2 => v3 and change a few trivial things. We don't have an owner for this though

**Comment by howardjohn** [comment]
not stale..

**Comment by howardjohn** [comment]
I think the migration path will look like this:

For people using revisions:
* Upgrade to 1.7, everything uses v3

For people not using revisions:
* Upgrade pilot to 1.7. Pilot understands v2 and v3 transport protocol. Pilot sends v3 config only. Immediately, every single cluster/listener in the entire mesh drains as config is reloaded for v3, causing huge load on the cluster.

The "people using revisions" group ends up in a good state, others not so great...

The other alternative is to completely support both API versions. This basically means 2x the pilot code, as everything is duplicated, which seems even worse

**Comment by howardjohn** [comment]
Actually we cannot send v3 clusters without a bootstrap config change. Therefor we must support v2+v3 simultaneous or enforce revisions for upgrade, as far as I can tell?

**Comment by howardjohn** [comment]
Dump of info from another discussion:

Tentative plan:

Istio 1.6 generates xDS v2, Envoy in 1.6 accepts xDS v2 and v3
Istio 1.7 generates xDS v3, Envoy in 1.7 accepts xDS v2 and v3
Istio 1.8 generates xDS v3, Envoy in 1.8 only accepts xDS v3

However, this does not support in place upgrades. This is because Envoy is the one that determines the api version to use (eg it requests "V3 clusters"). This means older proxies will request v2 to a 1.7 istiod, which will not know how to serve that version. Because of this, we will rely on multiple control planes to ensure we don't have this version mismatch.

Additional background info:

XDS versioning has two parts, the transport version (v2 and v3 is roughly the same currently) and the resource versions. They are independent of each other. Additionally we could serve just v3 endpoints and v2 cds/lds/eds if we wanted to. In practice, we would struggle if we do this because EDS and CDS share a lot of code, as do LDS and RDS. If we do need to delay the LDS v3 migration due to the bind_to_port stuff, we could just transition EDS and CDS over.

I think its feasible to support v2 and v3 transport in the same Istiod version due to the minimal amount of code there. However to support both resource versions would be a ton of work. I think Envoy has some internal conversion logic/automation but it would take substantial work to convert that to golang (and probably major performance impact). To duplicate all of our code over is not really sustainable in my opinion. Because of this, I think it makes sense to just do a hard shift over for 1.7

**Comment by howardjohn** [comment]
https://github.com/envoyproxy/envoy/issues/10776 might change our approach. Except if that change were to be merged but only landed in 1.15.0 that would mean we couldn't migrate to v3 in 1.7 as planned and would need to delay until 1.8if I am understanding correctly. Since the previous version needs to have the version mixing patch first in order for us to do the upgrade since we upgrade the control plane first - I guess we could upgrade the proxy first if we needed

**Comment by howardjohn** [comment]
Here are the options we discussed:

**1. Hard cut over to v3 in 1.7. To support upgrade without (control plane) downtime, users must use revisions.**

Pros:
* Lowest development cost, both upfront and in maintenance

Cons:
* Users must use revisions - in place upgrade is not supported.

**2. Serve v2 and v3 in Pilot in 1.7. In 1.8 we can drop v2.**

Pros:
* Most compatibility
* Is a gradual approach, which is generally safer. I think in this case its a very weak pro, as v2 and v3 are nearly identical so its low risk

Cons:
* Major tech debt - a large chunk pilot code base is effectively doubled. This means 2x testing, all changes require change 2 places, etc

**3. Implement support in Envoy for either mismatched request/response versions (https://github.com/envoyproxy/envoy/issues/10776) or fallback to a different version (https://github.com/envoyproxy/envoy/issues/8421#issuecomment-613619639). Pilot just serves v3.**

Pros:
* Lowest development cost, both upfront and in maintenance for Pilot. There will be development costs on Envoy side, but these may be done by others - this is a problem shared by all control planes.
* We can support in place upgrades

Cons:
* Unless the change is backported to Istio 1.6/Envoy 1.14, we will need to delay our upgrade to 1.8, as we will need the previous version's proxies to support this.

**4. Implement a conversion layer from v3 -> v2, or v2 -> v3.**

Pros:
* The core code base is in one version, so costs are lower.

Cons:
* This conversion may be extremely complicated, if not infeasible.
* Likely a large performance cost associated, especially if we have to remarshall Any types.

**Comment by howardjohn** [comment]
My personal preference would be 1 or 3. If we can backport 3 to Istio 1.6 so we aren't delayed that becomes even more compelling.

**Comment by howardjohn** [comment]
@costinm @nrjpoddar @htuch let me know if you have any other comments or opinions

**Comment by nrjpoddar** [comment]
@howardjohn you covered all the cases well with pros/cons. For me relying on upstream Envoy features has historically been a bottleneck so option 3 can be tricky. 

Option 1 or 2 seem to be the realistic choices afaict as we control the timeline.

Looks like if we do option 1 and option 3 happens to land even better for future but if we go with option 2 and option 3 becomes available a lot of work will possibly be for nothing.

**Comment by howardjohn** [comment]
If we do option 1, we would require option 3 to be backported to get the benefits though, right? I am not sure how feasible that is to backport, or even how likely it is to get implemented. We would need to defer to some Envoy maintainers though.

**Comment by htuch** [comment]
@howardjohn if you don't need to support both v2 and v3 clients at the same time, then you are much better off only supporting a single version, so from that perspective, your plan seems reasonable. I think the items in option 3 need some work to make happen, but https://github.com/envoyproxy/envoy/issues/10776 shouldn't be too hard.

If you need to support gRPC as well, you need to work within their API clock cycle.

**Comment by seflerZ** [comment]
@howardjohn Thanks for your effort! I'm diving into incremental xDS recently and found migrating to v3 is necessary before that. 

- First, I think option 2 is not practical with currently architecture because it involves much work and hard to implement and maintain.
- Second, for option 4, a convention layer is far beyond the context of Istio and it depends on the protocol to support convention. I do not recommend.
- 3rd. 1 looks like feasible approach as we can rely on multiple control planes as you've mentioned. But it is not a long term solution since if we'll move to v4 someday. The users must migrate their proxies again. If they have thousands of proxies, the process may not be easy and will cost much.

I think the reason why we're facing this problem today is we have proxy logic and Pilot core model really coupled and don't even have any abstract components about xDS(we put all the codes within v2 today). That is to say, it we decide to switch to a newer version like v4. We must copy the codes to namespace v4 and do refactoring entirely.

Generally speaking, the best practice to switch to a new version of protocol is with completely backward compatibility support. Such as from 4G to 5G in telecom industries, you can NOT tell users to throw their 4G phones anyway. The 4G stations still should work. The services should be there without any extra configuration.

Therefore, I think we need a architect solution and make it easier for next upgrading. We need some architecture refactoring to decouple the proxy and Pilot core model is good. Make it modular and plug-able.

(#10776)[https://github.com/envoyproxy/envoy/issues/10776] suggests it to send v3 resources for v2 requests. I'm not sure how the old proxies would be able to handle the new versions of resources.

@hzxuzhonghu @lizan How do you think about this?

**Comment by hzxuzhonghu** [comment]
Fair point, currently requires users to upgrade all their proxies is not acceptable for production users. The backward compatibility should be the first priority stuff to take into consideration.
@howardjohn @rshriram we have done much breaking the compatibility work before in order to reduce control plane complexity. But real users are not buying it. They only care why my old workloads do not work. About this we should learn from k8s, https://kubernetes.io/docs/reference/using-api/deprecation-policy/

**Comment by lizan** [comment]
Yeah that seems a fair point. We would need to upgrade to v4 so in long term 2 with some mixture of 3 makes sense. 

> Unless the change is backported to Istio 1.6/Envoy 1.14, we will need to delay our upgrade to 1.8, as we will need the previous version's proxies to support this. The impact of a delayed migration to v3 is that we will be stuck on a frozen version, so we cannot pick up any changes to Envoy for a whole release, and we won't be able to give any feedback to Envoy on v3.

At this point I think we would delay the full v3 API upgrade (at least for LDS) to 1.8 anyway, because the virtual listener / hitless filter chain upgrade didn't make it into Envoy 1.14. cc @lambdai

**Comment by htuch** [comment]
@lizan 1.8 is EOY AFAIU, which seems precariously late; any slippage and you won't be able to consume Envoy cut at EOY 2020.

@hzxuzhonghu @seflerZ I'm not sure if your feedback on backward compatibility is for Istio support or for xDS itself. If it's for xDS, please file an issue at https://github.com/envoyproxy/envoy/issues and we can discuss this further. So far, we have maintained backward compatibility over a sliding window, we need to understand more how users are interacting with this policy.

**Comment by howardjohn** [comment]
Everyone, just a note I added an option 5 to the list above. Credits to @costinm for this idea. Now that I have thought about it more I think this might be the best move, provided it works. Some people on the Envoy slack tried it out in their control planes and it seemed promising though.

---

@htuch 

> if you don't need to support both v2 and v3 clients at the same time, then you are much better off only supporting a single version, so from that perspective, your plan seems reasonable

We don't need to support older clients for extended periods of time, but we do need to support upgrades. When doing upgrades we bump the control plane first, so it needs to be able to work with the N-1 version. In this case the N-1 would be v2.

In Istio 1.6 we are adding a new feature of having multiple control plane versions running at the same time, so you would have v2 -> v2 and v3 -> v3. However its a new feature and currently not required, so there are some issues with making it required.

> If you need to support gRPC as well, you need to work within their API clock cycle.

I think this is a non-issue for us. We don't currently support gRPC, and when we do (there is some WIP in this area) the generation logic is currently independent of the Envoy logic, so we could have it on v2 if we really needed.

> I'm not sure if your feedback on backward compatibility is for Istio support or for xDS itself.

From what I can see its fairly easy to handle changes within the v2 api. As new fields are added, we rely on the allow_unknown_fields or we have version checks in the generation logic if we need to be smarter about it. The same goes for deprecations. What is harder about this one is its not a rolling window of changes going in, its a hard cut from `v2` to `v3`. However, if we do option (3) or (5) this would be just as simple as other updates. The only real change then would be the small refactoring needed for `v2` -> `v3` in all of the imports (which, by the way, is much more ergonomic in v3 - thanks whoever made those changes!).

---

@seflerZ 

> But it leaves the work to users and is not a long term solution. Since if we'll move to v4 someday, they must migrate their proxies again. If there are thousands of proxies, the process may not be that easy and will cost much.

I don't see why this is a concern? You already need to update all of your proxies for every version of Istio, why is it any different for xds versions?

> Such as from 4G to 5G in telecom industries, you can NOT tell users to throw their 4G phones anyway. 

An Envoy is not a phone though. I absolutely think we can tell a user to throw away their Envoy - that is how upgrades work.

> I'm not sure how the old proxies would be able to handle the new version resources (apply a patch before upgrading?).

The old proxies (1.5 and 1.6) already support v3 resources. We do not support more than N-1 proxy so its not really any issue for us?

> For example, we can move CDS to v3 with incremental supported first and leave others still in v2.

As mentioned above, I don't think its feasible to migrate just one resource. We should migrate CDS/EDS together at the least, and maybe LDS/RDS together due to the coupling of the two APIs. I had a prototype just moving CDS and it was almost 2x the code because I had to duplicate everything.

**Comment by nrjpoddar** [comment]
@howardjohn Option 5 and 2 only differ in implementation but are semantically the same. If we end up using a lot of deprecated (i.e. only v2) and new (i.e. only in v3) fields in any release, we end up at the same state unless as Option 2 I'm missing something.

**Comment by howardjohn** [comment]
Option 2 means copying 10k lines of code. Option 5 is just a few lines of code to do `if xdsV2 { response.TypeUrl = v2TypeUrl }` in a few places.

We do not use any deprecated fields except bind_to_port, which is a prerequisite for all of this work anyways. So the only thing we need to do is if we add any v3 only fields, we guard them with a version check - this is something we already need to do for new fields.

**Comment by nrjpoddar** [comment]
> Option 2 means copying 10k lines of code. Option 5 is just a few lines of code to do `if xdsV2 { response.TypeUrl = v2TypeUrl }` in a few places.
> 

Ah I never realized that you literally meant copying, I was always assuming if/else blocks for backwards compatibility.

> We do not use any deprecated fields except bind_to_port, which is a prerequisite for all of this work anyways. So the only thing we need to do is if we add any v3 only fields, we guard them with a version check - this is something we already need to do for new fields.

If we think it's low tech debt for us, let's do this!

**Comment by lizan** [comment]
> @lizan 1.8 is EOY AFAIU, which seems precariously late; any slippage and you won't be able to consume Envoy cut at EOY 2020.

@htuch I meant "v3 only" by full upgrade. v2/v3 are both supported in 1.7 and we drop v2 in 1.8.

**Comment by howardjohn** [comment]
Based on https://github.com/istio/istio/pull/23032 working pretty well, I suggest we move forward with this approach.

User impact:
* EnvoyFilters need to updated to use v3 instead of v2

Development impact:
* Minimal. Everything in pilot will be v3. We just need to make sure to version guard any v3 specific features we use.

Plan of action:
Once 1.6 is cut, we make the following changes. Order is not exact, I have decomposed it a bit to avoid a single mega-PR:
* Update the ads code to support swapping type URLs
* Update ADS code to support v2 and v3 transport code. This will require some duplication but its pretty minimal
* Update CDS/EDS to v3
* Update RDS/Listener filters to v3
* Update LDS to v3 (blocked by bind_to_port)
* Update bootstrap to start requesting v3/using v3 transport
* Once 1.7 ships, we can remove all v2 code

**Comment by hzxuzhonghu** [comment]
>>But it leaves the work to users and is not a long term solution. Since if we'll move to v4 someday, they must migrate their proxies again. If there are thousands of proxies, the process may not be that easy and will cost much.

>I don't see why this is a concern? You already need to update all of your proxies for every version of Istio, why is it any different for xds versions?

>>Such as from 4G to 5G in telecom industries, you can NOT tell users to throw their 4G phones anyway.

>An Envoy is not a phone though. I absolutely think we can tell a user to throw away their Envoy - that is how upgrades work.

@howardjohn I donot agree with this, like in GKE, let's say, we used to deploy a workload with extensions/v1beta1 deployment. However, if users upgrade the k8s version from previous 1.5 to 1.13, you can not require users to stop the workload and recreate a new apps/v1 deployment. 

Istio is a control plane, and similar to k8s api. If we require users to restart all their workloads each time they upgrade the controlplane, this is absolutely not aa good solution. It will largely influence it being used in production.

**Comment by howardjohn** [comment]
@hzxuzhonghu I think it's an interesting topic but this issue is already really complicated. Can we split off this conversation into another issue or proposal? Our current policy is N-1 support for proxy version, so until that changes we should implement the v3 migration with that assumption

**Comment by seflerZ** [comment]
> > But it leaves the work to users and is not a long term solution. Since if we'll move to v4 someday, they must migrate their proxies again. If there are thousands of proxies, the process may not be that easy and will cost much.

> I don't see why this is a concern? You already need to update all of your proxies for every version of Istio, why is it any different for xds versions?

My concern is every time we update xDS API the users need to replace their Envoy Sidecards. If the scale is 100~, it might be okay. How about 10k+ Envoy Sidecards? What if those sidecars belong to different Teams? Who is responsible for the direction? It may be a hard work.

Okay, for option 3. If Envoy handles mixed version response well (Such as requesting v2 and response is v3 and functions well), we can do in place upgrades with clean code. This is a problem of all control planes as mentioned.

I agree with @lizan, we support requesting v2/v3 in 1.7 and drop v2 in 1.8.

**Comment by howardjohn** [comment]
We have supported N-1 for a long time, nothing is changing with our support policy. And now we can have multiple control planes running at once so you can have old versions running if you need to - I just ran a cluster with Istio 0.2 alongside 1.6 without issues. As mentioned above this has been out support policy for quite a while so if you want to change that you should present a separate proposal.

**Comment by seflerZ** [comment]
John and I talked much. According to this [doc](https://github.com/envoyproxy/envoy/blob/bb7ceff4c3c5bd4555dff28b6e56d27f2f8be0a7/docs/root/configuration/overview/versioning.rst). The v3 objects may be transferred over the v2 transport protocol. So if we set typeURL to v2 to v3 objects, it can function well with Pilot has v3 code base only.

For attributes added in v3, I think there is no need to worry about. It's the attribute changes may need some effort to handle. We can use conditional logic. This wouldn't be a problem since we already do like that and will remove v2 in 1.8.

**Comment by lambdai** [comment]
> Update LDS to v3 (blocked by bind_to_port)

@howardjohn LDS v3 still support bind_to_port so it's not a hard blocker.

The legit config space is different, I prefer to leave bind_to_port as a backup path

**Comment by howardjohn** [comment]
@lambdai so we could, in theory, migrate to v3 today?

**Comment by htuch** [comment]
Uh yeah, we should not have allowed `bind_to_port` to make its way into v3, that is definitely not intended, CC @envoyproxy/api-shepherds

**Comment by howardjohn** [comment]
Another concern is that we need to be able to support EnvoyFilters acting on v2 config

**Comment by hzxuzhonghu** [comment]
@howardjohn @htuch what's the behavior for old proxies when they got v3 response, mostly concerned the newly added fields in v3 like `Listener.HiddenEnvoyDeprecatedUseOriginalDst`. Is it ignored?

**Comment by howardjohn** [comment]
Listener.HiddenEnvoyDeprecatedUseOriginalDst is not a new field it's just renamed, the proto wire format is identical. When an old proxy gets v3 it should be fine as long as it's not super old and doesn't support v3 at all which we don't support such a large gap

**Comment by hzxuzhonghu** [comment]
Then will envoy convert the v3 `HiddenEnvoyDeprecatedUseOriginalDst` field into v2  `UseOriginalDst`?

If not, how can we make sure it works as expected?

**Comment by howardjohn** [comment]
It doesn't "convert" it. It's literally the same field on the wire. There is not field name as part of the wire protocol, so it's exact same for v2 and v3 so no conversion necessary

**Comment by hzxuzhonghu** [comment]
IC, it is protobuf, not json

**Comment by howardjohn** [comment]
TODO: use `--bootstrap-version=v3`. Blocked by envoy updates

**Comment by esnible** [comment]
I have created https://github.com/istio/istio/pull/24516 to update `istioctl proxy-config listeners` for XDS v3.  I encourage folks from the Networking WG to review it.

I would like to make sure that the other `istioctl proxy-config` subcommands are working.  They seem to be.  I don't know a good way to compare `proxy-config` output with and without the PRs associated with this issue.  If anyone sees or expects a problem with `proxy-config` please mention it in #user-experience Slack.

**Comment by howardjohn** [comment]
This is complete
